Abandoned - Fixes #3691. Adds `Popover`

tig opened the pull request:
## Fixes

- Fixes #3691

![fykztjo 1](https://github.com/user-attachments/assets/02f31f30-6155-40e4-8d33-46bf8933edd5)

![fImBrr3 1](https://github.com/user-attachments/assets/003db3ad-d916-4119-9e85-6de08a0355e0)

- Fixes #3757

## Proposed Changes/Todos

- [x] Prototype in `View Experiments` the concept described in #3691 to prove concept.
- [x] Prototype in `Bars` to show how `ContextMenu` and `MenuBar` could utilize
- [ ] Re-implement `ContextMenu` using `MenuV2` - Initially no cascading menus for simplicity
- [ ] Re-implement `AutocompletePopup`
- [ ] Re-implement `Combobox`
- [ ] Add more ShadowView unit tests (see #3757)

## Future PRs

- [ ] Replace `Menu` and `MenuBar` using `Bar` (aka finish `Menuv2`

## Pull Request checklist:

- [x] I've named my PR in the form of "Fixes #issue. Terse description."
- [x] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [x] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [ ] I conducted basic QA to assure all features are working

tig commented:
FWIW, here's the API in action (the prototype for the `File` menubar item in `Bars`):

```cs
var fileMenuBarItem = new Shortcut
{
    Title = "_File",
    HelpText = "File Menu",
    Key = Key.D0.WithAlt,
};

fileMenuBarItem.Accept += (sender, args) =>
                          {
                              var fileMenu = new Menuv2
                              {
                                  Id = "fileMenu",
                              };
                              ConfigureMenu (fileMenu);
                              fileMenu.VisibleChanged += (sender, args) =>
                                                         {
                                                             if (Application.Popover is { Visible: false })
                                                             {
                                                                 Application.Popover?.Dispose ();
                                                                 Application.Popover = null;
                                                             }
                                                         };
                              Application.Popover = fileMenu;
                              Rectangle screen = fileMenuBarItem.FrameToScreen ();
                              fileMenu.X = screen.X;
                              fileMenu.Y = screen.Y + screen.Height;
                              fileMenu.Visible = true;

                          };
```

Here's the entirety of the new API (on `Application`):

```cs
    /// <summary>Gets or sets the Application Popover View.</summary>
    /// <remarks>
    ///     <para>
    ///         To show or hide the Popover, set it's <see cref="View.Visible"/> property.
    ///     </para>
    /// </remarks>
    public static View? Popover
```

tig commented:
![ZtNvfaH 1](https://github.com/user-attachments/assets/f05b54d9-9edf-4de3-8f0f-d87d59510cdd)

tig commented:
Kinda cool. Mouse wheel will now be able to nav within a Bar (or Menu).

![AKE5pyM 1](https://github.com/user-attachments/assets/cd080c64-7a83-4128-a547-46b5f4ac4e60)

tig commented:
![n4wBc8q 1](https://github.com/user-attachments/assets/53b44128-7d37-4828-91a1-293136fb15b9)

A new `ContextMenuv2` implementation. 

`TextField` now uses it instead of the old one.

BDisp commented:
> A new `ContextMenuv2` implementation.
> 
> `TextField` now uses it instead of the old one.

But the start position isn't the same as the old one. Please pay attention to that detail.

tig commented:
Will do. Just work in progress for now.

tig commented:
> > A new `ContextMenuv2` implementation.
> > `TextField` now uses it instead of the old one.
> 
> But the start position isn't the same as the old one. Please pay attention to that detail.

@BDisp Can you explain the logic you used for the old design?

I implemented what's in this PR without actually looking at the old behavior. I implemented what I thought made sense. Here's my thinking, for `TextField`:

- The ContextMenu should be 1 row below or above the TextView. Below if it will fit below. Above if it will not fit below. 
  - If it won't fit either, then it can appear over the TextView. 
- The left side of the ContextMenu should line up with the cursor position in the TextField. This makes sense because the actions like Paste impact the cursor position. If the menu won't fit such that this works, it will be positioned such that it fits. 

I now see, the old behavior didn't work like this. The old behavior for `TextView` is:

- The ContextMenu should be 1 row below or above the TextView. Below if it will fit below. Above if it will not fit below. 
  - If it won't fit either, then it can appear over the TextView. 
- The left side of the ContextMenu should line up with left side of TextField.  If there's not room it can be positioned so it fits.

For an arbirary View, with the mouse, in this PR:

- The top-left corner of the ContextMenu appears where the mouse is, unless it can't fit on the screen. Then it will appear such that it is fully visible.

The old behavior:

- The first cell of the first menuitem is positioned where the mouse is.

I think the old behavior in this case is correct and will change this PR to match that.

What are your thoughts on the above?

BDisp commented:
The view  that use `ContextMenu` can customize his position if he want. The `TextField` has no sense to start his position at cursor position, `TextView` has. `TextField` must start his position at (0,1), unless it doesn't fit but never over it. If doesn't fit below put it above the top and not over it. In `TextView` always put it at cursor position if it fits, otherwise put it above the line where the cursor is positioned, but never over it, unless no screen space is available.

tig commented:
> The TextField has no sense to start his position at cursor position, TextView has. TextField must start his position at (0,1), but never over it unless it doesn't fit.

I don't understand. 

TextView does know the cursor pos. I've used it to position the CM in this PR. I'm travelling today so can't send a video. Try it and let me know what you think.

BDisp commented:
Nothing of this is working in your PR. You put a zig zag menus for what? To show that this will stopping working? `UseSubMenusSingleFrame` will also not working in the new Popover? To me a `ContextMenu` must work like a `Menu` but without a visible `MenuBar`. `Popover` should be used for floated view that will show content on front of others views. But that only my opinion.

![image](https://github.com/user-attachments/assets/59da9e17-f963-4b36-809f-a3bd788cf999)

BDisp commented:
In this PR:

![WindowsTerminal_sfT355oQXG](https://github.com/user-attachments/assets/de3d565c-7171-467a-a717-329653f1d0e0)

In v2_develop the `MenuBar` is not showing.

![WindowsTerminal_TFdFrY5i9k](https://github.com/user-attachments/assets/25f34d60-7c73-49bd-aa82-ff6e7a83753f)

tig commented:
> Nothing of this is working in your PR. You put a zig zag menus for what? To show that this will stopping working? `UseSubMenusSingleFrame` will also not working in the new Popover? To me a `ContextMenu` must work like a `Menu` but without a visible `MenuBar`. `Popover` should be used for floated view that will show content on front of others views. But that only my opinion.
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/59da9e17-f963-4b36-809f-a3bd788cf999)
> 
> 

This is a WIP. I will add the more advanced features as I progress.

tig commented:
> In this PR:
> 
> 
> 
> ![WindowsTerminal_sfT355oQXG](https://github.com/user-attachments/assets/de3d565c-7171-467a-a717-329653f1d0e0)
> 
> 
> 
> In v2_develop the `MenuBar` is not showing.
> 
> 
> 
> ![WindowsTerminal_TFdFrY5i9k](https://github.com/user-attachments/assets/25f34d60-7c73-49bd-aa82-ff6e7a83753f)
> 
> 
> 
> 

This is because Window is set for Overlapped/Movable/Sizable. 

Please file a bug for me to make it so arrange mode is disabled if Top.

tig commented:
Here's how this PR current works:

There's a new property on `Application`:

```cs
    /// <summary>Gets or sets the Application Popover View.</summary>
    /// <remarks>
    ///     <para>
    ///         To show or hide the Popover, set it's <see cref="View.Visible"/> property.
    ///     </para>
    /// </remarks>
    public static View? Popover
```

When set, if there's an Popover it's made `Visible = false`. Then the new View is set as the popover, initialized if needed.

A handler for `_popover.VisibleChanged` is added which does the heavy lifting.

To show a Popover, e.g. on a right mouse click, simply :

- Set your View's X/Y to the Screen-relative coords you want
- Set `Application.Popover` to your View
- Set `yourView.Visible = true`

```cs
        void TestFrameOnMouseClick (object sender, MouseEventArgs e)
        {
            if (e.Flags == MouseFlags.Button3Clicked)
            {
                popoverView.X = e.ScreenPosition.X;
                popoverView.Y = e.ScreenPosition.Y;
                Application.Popover = popoverView;
                Application.Popover!.Visible = true;
            }
        }
```

The View will remain visible and will prevent any other view from getting mouse/keyboard input until 

- The user clicks outside of the popover's fame
- The user presses QuitKey
- Someone sets `Applicaiton.Popover.Visible = false`
- Someone sets a new View to `Application.Popover`

This works fairly brilliantly for:

## Standalone use-cases as shown in `ViewExperiments.cs`
![06cV1tE 1](https://github.com/user-attachments/assets/de5c9d94-edca-4878-bf01-f5c556ef3541)
S
## ContextMenu (hacked to use `Menuv2/Bar` - No sub-menus
![Q7F4sIY 1](https://github.com/user-attachments/assets/4f1c0048-811f-4bba-8d04-23e10b07a529)

## Menuv (Prototype)
![TXHKxsX 1](https://github.com/user-attachments/assets/2ba8d85c-c5b1-4dd1-b490-116b005cab58)

# Problems

* It's really not a good design to couple more `View` logic in `Application`. I would really prefer that `Application` just knows about a single `Top`. This design actually works against what I'm trying to do with
  - #2491 

* `MenuBar` scenarios require the MenuBar, which is not a Popover, to still receive input when a sub-menu is shown as the Popo er. All modern menu systems, esp on Windows, let you hover over the menu bar once activated and sub-menus are automatcially shown/hidden as you do so:

![Ifu2n1N 1](https://github.com/user-attachments/assets/3a994ee0-deb3-4c04-9704-706769bbc6fa)

![qy3du96 1](https://github.com/user-attachments/assets/eb60dc9a-f97e-46b4-83d2-4f795cbb0a7b)

With this design, there's no clean way to make this happen. Here's what it looks like now (I have to click on a menu bar item to activate it's submenu):
![NfYHxD5 1](https://github.com/user-attachments/assets/53b69914-6766-49e5-aeaf-e6fcf9fdc410)

There's also no good way, with this design, to enable sub-menus. 

I've spent a bunch of time noodling on this particular issue and tried several ideas, all of which seem like hacks and just cause more issues:

- Transform `this._menuBar` into a `Popover` dynamically (e.g.. clone it and set the clone to `Application.Popover`). 
- Make `Application.Popover` be `Stack<View> Popovers` in conjunction with the clone idea.

So, for this reason, for now, I'm going to abandon this PR, fork the branch, and create a new PR that will try a different approach:

- Add `ViewArrangement.Popover` 
- Any view with this flag set, will be "homed" from a layout/mouse/keyboard input perspective as though they were subviews of `Application.Top`
